New Astro Runtime Image Doc (#785)
astronomer/docs: software/runtime-image-architecture.md

---
sidebar_label: 'Astro Runtime Architecture'
title: 'Astro Runtime Architecture'
id: runtime-image-architecture
description: Reference documentation for Astro Runtime, a differentiated distribution of Apache Airflow.
---

Astro Runtime is a production ready, data orchestration tool based on Apache Airflow that is distributed as a Docker image and it is required by all Astronomer products. It is intended to provide organizations improved functionality, reliability, efficiency, and performance.

If your organization is using Software version 0.29 or later, you can deploy Astro Runtime images. Astro Runtime includes the following features:

> - The `astronomer-providers` package. This package is an open source collection of Apache Airflow providers and modules that is maintained by Astronomer. It includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`.

Reviewer: Update this list with Astro doc changes? Copy it over and we might need to subtract/replace docs links but I think it's mostly 1:1

```

## Python Versioning

Astro Runtime supports Python package version 3.9.

## Astro Runtime and Apache Airflow Parity

This table lists Astro Runtime releases and their associated Apache Airflow versions.

| Astro Runtime                                   | Apache Airflow Version                                                                | 
| ----------------------------------------------- | ----------------------------------------------------------------------------- |
| 3.0.x                                       |       2.1.1                                                                       |
| 4.0.x                                       |       2.2.0                                                                    | 
| 4.1.x                                       |       2.2.4                                                                    |
| 4.2.x                                       |       2.2.4                                                                      | 
| 5.0.x                                       |       2.3.0                                                                        | 

> ## System Distribution

Reviewer 2: This needs to be updated per the comment in the Astro doc.

Author: @jwitz sorry, I'm not clear what edit is required here.

---
sidebar_label: 'Astro Runtime Architecture'
title: 'Astro Runtime Architecture'
id: runtime-image-architecture
description: Reference documentation for Astro Runtime, a differentiated distribution of Apache Airflow.
---

Astro Runtime is a production ready, data orchestration tool based on Apache Airflow that is distributed as a Docker image and required by all Astronomer products. It is intended to provide organizations improved functionality, reliability, efficiency, and performance.

If your organization is using Software version 0.29 or later, you can deploy Astro Runtime images. Astro Runtime includes the following features:

- The `astronomer-providers` package. This package is an open source collection of Apache Airflow providers and modules that is maintained by Astronomer. It includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`.
- Airflow UI improvements. For example, showing the Deployment Docker image tag in the footer of all UI pages.
- Exclusive features such as new Airflow components and improvements to the DAG development experience.

> For more information about the features that are available in Astro Runtime releases, see the [Astro Runtime Release Notes](/astro/runtime-release-notes).

Reviewer 2: ```suggestion
For more information about the features that are available in Astro Runtime releases, see the [Astro Runtime Release Notes](https://docs.astronomer.io/astro/runtime-release-notes).
```
When you cross link to an Astro doc, you have to use the entire link. Our site considers these two docsets to be in their own uncrossable universes, so relative linking doesn't work like expected.

The Dockerfiles for all supported Astro Runtime images are located in the [Astronomer `astro-runtime` GitHub repository](https://github.com/astronomer/astro-runtime).

## Executors

In Airflow, the executor is responsible for determining how and where a task is completed. On Astronomer Software, Astro Runtime supports the following executors:

- [Local Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/local.html)
- [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html)
- [Kubernetes Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/kubernetes.html)

Soon, Astronomer will provide a new executor with intelligent worker packing, task-level resource requests, improved logging, and Kubernetes-like task isolation.

## Provider Packages

To make it easy to get started with Airflow, all Astro Runtime images have the following set of open source provider packages pre-installed:

> - Amazon (apache-airflow-providers-amazon)
> - Elasticsearch (apache-airflow-providers-elasticsearch)
> - Celery (apache-airflow-providers-celery)
> - Google (apache-airflow-providers-google)
> - Password (apache-airflow-password)
> - Cloud Native Computing Foundation (CNCF) Kubernetes (apache-airflow[cncf.kubernetes)
> - PostgreSQL (Postgres) (apache-airflow-providers-postgres)
> - Redis (apache-airflow-providers-redis)
> - StatsD (apache-airflow-statsd)
> - Virtualenv (apache-airflow-virtualenv)
> - OpenLineage with Airflow (openlineage-airflow)
> - Astronomer Providers (astronomer-providers)

Reviewer 2: ```suggestion
- Amazon (`apache-airflow-providers-amazon`)
- Elasticsearch (`apache-airflow-providers-elasticsearch`)
- Celery (`apache-airflow-providers-celery`)
- Google (`apache-airflow-providers-google`)
- Password (`apache-airflow-password`)
- Cloud Native Computing Foundation (CNCF) Kubernetes (`apache-airflow-cncf.kubernetes`)
- PostgreSQL (Postgres) (`apache-airflow-providers-postgres`)
- Redis (`apache-airflow-providers-redis`)
- StatsD (`apache-airflow-statsd`)
- Virtualenv (`apache-airflow-virtualenv`)
- OpenLineage with Airflow (`openlineage-airflow`)
- Astronomer Providers (`astronomer-providers`)
```

Reviewer 2: All of these packages have links on PyPi as well, up to you whether we should hyperlink those. For example" https://pypi.org/project/openlineage-airflow/

Author: @jwitz links added.

> If an Astro Runtime release includes changes to an installed version of a provider package that is maintained by Astronomer (`astronomer-providers` or `openlineage-airflow`), the version change is documented in the [Astro Runtime Release Notes](/astro/runtime-release-notes.md).

Reviewer 2: ```suggestion
### Provider Package Versioning

If an Astro Runtime release includes changes to an installed version of a provider package that is maintained by Astronomer (`astronomer-providers` or `openlineage-airflow`), the version change is documented in the [Astro Runtime Release Notes](/astro/runtime-release-notes.md).
```

To determine the version of a provider package installed in your current Astro Runtime image, run:

```
docker run --rm {image} pip freeze | grep apache-airflow-provider
```
OpenLineage with Airflow standardizes the definition of data lineage, the metadata that forms lineage data, and how data lineage data is collected from external systems. OpenLineage with Airflow is intended to make it easier for organizations to integrate their lineage tools with Apache Airflow. See [OpenLineage and Airflow](https://docs.astronomer.io/astro/data-lineage-concepts#openlineage-and-airflow).

> The `astronomer-providers` package is a collection of Apache Airflow OSS providers and modules. The package is installed on Astro Runtime by default and is maintained by Astronomer. The package includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`. See [Deferrable Operators](https://docs.astronomer.io/astro/deferrable-operators). To access the source code for this package, see the [Astronomer Providers GitHub repository](https://github.com/astronomer/astronomer-providers).

Reviewer 2: Let's move this information to before the versioning information

Author: @jwitz done.

> ## Python Versioning

Reviewer 2: I'm not sure whether it's worth maintaining the individual patch versions of Python in this table. Python patch versions usually only contain bug fixes, and users are much more worried about minor/ major versions that have breaking changes. We should just have a single sentence here saying "All Astro Runtime images are compatible only with Python 3.9."

Author: @jwitz Done.

This table lists the Python packages supported by Astro Runtime. You can't modify the Python versions.

| Astro Runtime                                   | Python Package                                                                | 
| ----------------------------------------------- | ----------------------------------------------------------------------------- |

> | **3.0.x**                                       |      **3.9.6**                                                                        |

Reviewer 2: Any reason we're bolding the version numbers? I'm cool with no styling, since there's nothing we need to distinguish from here.

Author: @jwitz this was the format in an existing doc. I thought this was the standard.

| **4.0.x**                                       |      **3.9.7**                                                                         | 
| **4.1.x**                                       |      **3.9.7**                                                                      |
| **4.2.x**                                       |      **3.9.7**                                                                     | 
| **5.0.x**                                       |      **3.9.7**                                                                  | 

## Astro Runtime and Apache Airflow Parity

This table lists Astro Runtime releases and their associated Apache Airflow versions.

| Astro Runtime                                   | Apache Airflow Version                                                                | 
| ----------------------------------------------- | ----------------------------------------------------------------------------- |

> | **3.0.x**                                       |       **2.1.1**                                                                       |
> | **4.0.x**                                       |       **2.2.0**                                                                    | 
> | **4.1.x**                                       |       **2.2.4**                                                                    |
> | **4.2.x**                                       |       **2.2.4**                                                                      | 
> | **5.0.x**                                       |       **2.3.0**                                                                        | 

Reviewer 2: ```suggestion
| **3.0.x**                                       |       **2.1.x**                                                                       |
| **4.0.x**                                       |       **2.2.x**                                                                    | 
| **4.1.x**                                       |       **2.2.x**                                                                    |
| **4.2.x**                                       |       **2.2.x**                                                                      | 
| **5.0.x**                                       |       **2.3.x**                                                                        | 
```
I'd also be fine with removing the patch references altogether

---
sidebar_label: 'Astro Runtime Architecture'
title: 'Astro Image Architecture'
id: runtime-image-architecture
description: Reference documentation for Astro Runtime, Astronomer Software's Docker image for Apache Airflow.
---

Astro Runtime is a discrete, production ready, data orchestration tool based on Apache Airflow. The Astro Runtime image is intended to provide organizations improved functionality, reliability, efficiency, and performance.

If your organization is using Software version 0.29 or later, you can use Astro Runtime images in your Software Deployments. Astro Runtime images include the following features:

> - The `astronomer-providers` package. This package is an open source collection of Apache Airflow providers and modules that is maintained by Astronomer. It includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`.

Author: I don't think a `## Features` heading is helpful or necessary. They are often viewed as marketing content and skipped. It's better to leave it in the overview in the hope it is read.

Astro Runtime Docker images are hosted on the Astronomer Docker registry and enable Airflow on Astro. All Astro projects require that you specify an Astro Runtime image in your Dockerfile, and all Deployments on Astro must run only one version of Astro Runtime. Every version of Astro Runtime correlates to an Apache Airflow version. You can run different versions of Astro Runtime on different Deployments within a given Workspace or Cluster.

The Astro Runtime Docker image incorporates additional functionality that makes it easier for organizations to get started with Airflow implementations. This includes:

- A robust testing suite that includes performance benchmarking and end-to-end functionality and upgrade testing.
- A built-in directory of sample DAGs that fully utilize Airflow functionality.
- A collection of pre-installed Airflow provider packages.

> - Compatibility with the Astronomer Platform.

Author: I have removed the content.

The Dockerfiles for all supported Astro Runtime images are located in the [Astronomer `astro-runtime` GitHub repository](https://github.com/astronomer/astro-runtime).

## Executors

Executors are the mechanism used to run task instances. Astro Runtime supports the [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html).

Soon, Astronomer will provide a new executor with intelligent worker packing, task-level resource requests, improved logging, and Kubernetes-like task isolation.

## Provider Packages

When you install the Astro Runtime image, the supported Apache Airflow open-source software (OSS) provider packages are included. The following are the provider packages Astro Runtime supports:

- Amazon
- Elasticsearch
- Celery
- Google
- Password
- Cloud Native Computing Foundation (CNCF) Kubernetes
- PostgreSQL (Postgres)
- Redis
- StatsD
- Virtualenv
- OpenLineage with Airflow

> - astronomer-providers

Author: The changes have been implemented.

> To determine the supported provider package version for a specific Astronomer Software release, see the release notes. Run the following command to determine the provider package versions that are supported by your current Astro Runtime image:

Reviewer: ```suggestion
If a release of Astro Runtime includes a change to the installed version of a provider package that is maintained by Astronomer (`astronomer-providers` or `openlineage-airflow`), that version change will be reflected in [Astro Runtime Release Notes](runtime-release-notes.md).

To determine the version of a provider package installed in your current Astro Runtime image, run:
```

I'd like to not make the "To determine the supported provider package version" statement an explicit call to action, because:

- I don't want users to filter through release notes to find versions
- We don't actually update versions in there for ALL provider packages -- just for `astronomer-providers` and `openlineage-airflow` because we maintain those
- I think this statement is informational, not actionable

Let's also make sure to actually link to release notes here

> To determine the supported provider package version for a specific Astronomer Software release, see the release notes. Run the following command to determine the provider package versions that are supported by your current Astro Runtime image:

Author: @paolaperaza I don't think a separate section is required. This information is in a logical location and is easy to locate. It follows related information.

```
docker run --rm {image} pip freeze | grep apache-airflow-provider
```
OpenLineage with Airflow standardizes the definition of data lineage, the metadata that forms lineage data, and how data lineage data is collected from external systems. OpenLineage with Airflow is intended to make it easier for organizations to integrate their lineage tools with Apache Airflow. See [OpenLineage and Airflow](https://docs.astronomer.io/astro/data-lineage-concepts#openlineage-and-airflow).

The `astronomer-providers` package is a collection of Apache Airflow OSS providers and modules. The package is installed on Astro Runtime by default and is maintained by Astronomer. The package includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`. See [Deferrable Operators](https://docs.astronomer.io/astro/deferrable-operators). To access the source code for this package, see the [Astronomer Providers GitHub repository](https://github.com/astronomer/astronomer-providers).

> ## Python Packages

Reviewer: ```suggestion
## Python Versioning
```

This is about which version of Python is supported by which version of Astro Runtime -- nothing to do with "packages". Given that it's a versioning matrix, what are your thoughts on consolidating this table with the `Astro Runtime and Apache Airflow Parity` below?

Author: @paolaperaza it's better to keep this information seperate to make it easier for users to locate quickly.

## Astro Runtime and Apache Airflow Parity

This table lists Astro Runtime releases and their associated Apache Airflow versions.

| Astro Runtime                                   | Apache Airflow Version                                                                | 
| ----------------------------------------------- | ----------------------------------------------------------------------------- |
| **3.0.x**                                       |       **2.1.1**                                                                       |
| **4.0.x**                                       |       **2.2.0**                                                                    | 
| **4.1.x**                                       |       **2.2.4**                                                                    |
| **4.2.x**                                       |       **2.2.4**                                                                      | 
| **5.0.x**                                       |       **2.3.0**                                                                        | 

> ## System Dependencies

Reviewer: This isn't really "System Dependencies" (i.e. packages) -- this is more `OS Distribution` that this image is built with. It will always be a type and version. Thoughts on consolidating this with the table above?

Similar to: https://docs.astronomer.io/software/version-compatibility-reference#astronomer-certified

Author: @paolaperaza I prefer keeping this information separate.
